src/lib/salad/bps_tree.h

```diff
@@ -1111,6 +1127,62 @@ bps_tree_debug_check_internal_functions(bool assertme);
 /* Same as BPS_TREE_MEMMOVE but takes count of values instead of memory size */
 #define BPS_TREE_DATAMOVE(dst, src, num, dst_bck, src_bck) \
 	BPS_TREE_MEMMOVE(dst, src, (num) * sizeof((dst)[0]), dst_bck, src_bck)
+/* Memmoves elements from one block to another. */
+#define BPS_TREE_DATAMOVE_ELEMS(dst_i, src_i, num, dst_blk, src_blk) \
+	BPS_TREE_DATAMOVE((dst_blk)->elems + (dst_i), \
+			  (src_blk)->elems + (src_i), \
+			  num, (dst_blk), (src_blk))
+/* Memmoves child_ids from one block to another. */
+#define BPS_TREE_DATAMOVE_CHILD_IDS(dst_i, src_i, num, dst_blk, src_blk) \
+	BPS_TREE_DATAMOVE((dst_blk)->child_ids + (dst_i), \
+			  (src_blk)->child_ids + (src_i), \
+			  num, (dst_blk), (src_blk))
+/* Memmoves child_cards from one block to another if the field is enabled. */
+#ifdef BPS_BLOCK_CHILD_CARDS_ARRAY
+#define BPS_TREE_DATAMOVE_CHILD_CARDS(dst_i, src_i, num, dst_blk, src_blk) \
+	BPS_TREE_DATAMOVE((dst_blk)->child_cards + (dst_i), \
+			  (src_blk)->child_cards + (src_i), \
+			  num, (dst_blk), (src_blk))
+#else
+#define BPS_TREE_DATAMOVE_CHILD_CARDS(dst_i, src_i, num, dst_blk, src_blk)
+#endif
+
+/* Memmoves available data from one block to another (elems, child_ids, etc.) */
+#define BPS_TREE_DATAMOVE_INNER(dst_blk, dst_i, src_blk, src_i, num) do { \
+	/* Note `num - 1` for elems, see the struct bps_inner definition. */ \
+	BPS_TREE_DATAMOVE_ELEMS(dst_i, src_i, (num) - 1, dst_blk, src_blk); \
+	BPS_TREE_DATAMOVE_CHILD_IDS(dst_i, src_i, num, dst_blk, src_blk); \
+	BPS_TREE_DATAMOVE_CHILD_CARDS(dst_i, src_i, num, dst_blk, src_blk); \
+} while (0)
+
+#ifdef BPS_BLOCK_CHILD_CARDS_ARRAY
+#define BPS_BLOCK_INFO(block_id, block_card) { \
+	(bps_tree_block_id_t)block_id, \
+	(bps_tree_block_card_t)block_card \
+}
+#else
+#define BPS_BLOCK_INFO(block_id, unused) { (bps_tree_block_id_t)block_id }
+#endif
+
+#ifdef BPS_BLOCK_CHILD_CARDS_ARRAY
+#define BPS_TREE_CARD_UP_LEAF(...) bps_tree_card_up_leaf_impl(__VA_ARGS__)
+#define BPS_TREE_CARD_UP_INNER(...) bps_tree_card_up_inner_impl(__VA_ARGS__)
+#else
+#define BPS_TREE_CARD_UP_LEAF(...)
+#define BPS_TREE_CARD_UP_INNER(...)
+#endif
+
+/**
+ * The config-dependent block information.
+ */
+struct bps_block_info {
```

Comment:
Passing this struct around looks ugly. Please consider storing own cardinality in each block (along with the array of child cardinalities). The increase in memory consumption should be negligible but then you wouldn't need to pass the cardinality around and introduce the `block_info` struct.

Reply:
The thing is, we won't need the children cardinality array if we introduce the inner block own cardinality. It will in fact have a better performance. See the benchmarks above.